### ISSUE TITLE
ReadOnly Attributes should not be included in parameter flattening

### ISSUE THREAD

**Issue by annatisch** [Azure/autorest#1534]
Hello,

I'm not sure if other generated SDK's use parameter flattening, but in Python we use it to a threshold of 2, and it's caused some difficulties in the KeyVault client, where one of the Body Parameters includes a read-only attribute.

Read Only attributes should not be included in the count towards flattening threshold, nor should it be exposed as a parameter on the method which currently seems to be the case.

The issue seems to be here:
https://github.com/Azure/autorest/blob/master/src/core/AutoRest.Extensions/SwaggerExtensions.cs#L403

My fix would be:

```
if (bodyParameterType != null && (bodyParameterType.ComposedProperties.Count(p => !p.IsConstant  && !p.IsReadOnly) <= settings.PayloadFlatteningThreshold || bodyParameter.ShouldBeFlattened()))
{
    var parameterTransformation = new ParameterTransformation
    {
        OutputParameter = bodyParameter
    };
    method.InputParameterTransformation.Add(parameterTransformation);

    foreach (var property in bodyParameterType.ComposedProperties.Where(p => !p.IsConstant && p.Name != null  && !p.IsReadOnly))
    {
        var newMethodParameter = new Parameter();
        newMethodParameter.LoadFrom(property);
        ...
```

Let me know if you think this looks reasonable and I'll submit a PR.
Cheers,
Anna

FYI @lmazuel, @tbombach

**Comment by annatisch** [comment]
@fearthecowboy - would this patch have any conflicts with your refactor?

**Comment by fearthecowboy** [comment]
A smidgen... I can make the change in here too. It's just changing the `!p.IsConstant `  to `!p.IsConstant  && !p.IsReadOnly` on those two lines?

**Comment by annatisch** [comment]
Yup, that's all it is - I think ReadOnly and Constant should be treated the same in this scenario - and no one has corrected me yet ;)

**Comment by fearthecowboy** [comment]
I've put this change into the `RefactorCodeModel` branch already, and since it'll be `master` pretty quick, don't worry about doing anything with the current master.

**Comment by annatisch** [comment]
Awesome - thank you!

**Comment by annatisch** [comment]
@fearthecowboy - it looks like this patch somehow got lost in the merge to master.
So I've included it in a PR: #1572
